File: backend/app/schemas/commission.py
# app/schemas/commission.py
"""Pydantic schemas for the commission module."""
from __future__ import annotations

from datetime import date, datetime
from decimal import Decimal
from typing import Any, Dict, List, Optional
from uuid import UUID

from pydantic import BaseModel, Field, EmailStr, field_validator

from app.services.sms_service import normalize_e164, is_valid_e164


def _validate_optional_phone(v: Optional[str]) -> Optional[str]:
    if v is None:
        return None
    raw = (v or "").strip()
    if not raw:
        return None
    normalized = normalize_e164(raw)
    if not normalized or not is_valid_e164(normalized):
        raise ValueError("Enter a valid phone number with country code (e.g. +919876543210)")
    return normalized


def _normalize_optional_email(v: Optional[str]) -> Optional[str]:
    if v is None:
        return None
    raw = str(v).strip()
    return raw or None


# ─── Payee ────────────────────────────────────────────────────────────────────

class CommissionPayeeCreate(BaseModel):
    display_name: str
    code: Optional[str] = None
    phone: Optional[str] = None
    email: Optional[EmailStr] = None
    external_user_id: Optional[str] = None
    link_type: str = "external"  # vendor_user | supplier | customer | external
    vendor_user_id: Optional[UUID] = None
    supplier_id: Optional[UUID] = None
    customer_id: Optional[UUID] = None
    default_payout_method: str = "bank_transfer"
    # bank_source: 'master' = read from linked master record, 'custom' = fields below
    bank_source: str = "master"
    bank_name: Optional[str] = None
    account_number: Optional[str] = None
    account_holder_name: Optional[str] = None
    ifsc_code: Optional[str] = None
    upi_id: Optional[str] = None
    wallet_provider: Optional[str] = None
    wallet_id: Optional[str] = None
    currency: str = "INR"
    status: str = "active"

    @field_validator("phone", mode="before")
    @classmethod
    def validate_phone(cls, v: Optional[str]) -> Optional[str]:
        return _validate_optional_phone(v)

    @field_validator("email", mode="before")
    @classmethod
    def validate_email(cls, v: Optional[str]) -> Optional[str]:
        return _normalize_optional_email(v)


class CommissionPayeeUpdate(BaseModel):
    display_name: Optional[str] = None
    code: Optional[str] = None
    phone: Optional[str] = None
    email: Optional[EmailStr] = None
    external_user_id: Optional[str] = None
    default_payout_method: Optional[str] = None
    bank_source: Optional[str] = None
    bank_name: Optional[str] = None
    account_number: Optional[str] = None
    account_holder_name: Optional[str] = None
    ifsc_code: Optional[str] = None
    upi_id: Optional[str] = None
    wallet_provider: Optional[str] = None
    wallet_id: Optional[str] = None
    currency: Optional[str] = None
    status: Optional[str] = None

    @field_validator("phone", mode="before")
    @classmethod
    def validate_phone(cls, v: Optional[str]) -> Optional[str]:
        return _validate_optional_phone(v)

    @field_validator("email", mode="before")
    @classmethod
    def validate_email(cls, v: Optional[str]) -> Optional[str]:
        return _normalize_optional_email(v)


# ─── Plan ─────────────────────────────────────────────────────────────────────

class CommissionPlanCreate(BaseModel):
    code: str
    name: str
    description: Optional[str] = None
    status: str = "active"
    effective_from: Optional[date] = None
    effective_to: Optional[date] = None
    payee_scope: str = "any"
    priority: int = 10
    stackable: bool = False
    settings: Dict[str, Any] = {}


class CommissionPlanUpdate(BaseModel):
    code: Optional[str] = None
    name: Optional[str] = None
    description: Optional[str] = None
    status: Optional[str] = None
    effective_from: Optional[date] = None
    effective_to: Optional[date] = None
    payee_scope: Optional[str] = None
    priority: Optional[int] = None
    stackable: Optional[bool] = None
    settings: Optional[Dict[str, Any]] = None

    @field_validator("code", mode="before")
    @classmethod
    def normalize_code(cls, v: Optional[str]) -> Optional[str]:
        if v is None:
            return None
        return str(v).strip() or None


# ─── Rule ─────────────────────────────────────────────────────────────────────

class CommissionRuleCreate(BaseModel):
    name: Optional[str] = None
    priority: int = 10
    is_active: bool = True
    applies_to: str = "all"
    product_id: Optional[UUID] = None
    service_id: Optional[UUID] = None
    category_id: Optional[UUID] = None
    uom: Optional[str] = None
    store_id: Optional[UUID] = None
    customer_group: Optional[str] = None
    channel: str = "any"
    event_tag: Optional[str] = None
    team_id: Optional[UUID] = None
    min_qty: Optional[Decimal] = None
    min_amount: Optional[Decimal] = None
    window_type: str = "per_line"
    period: Optional[str] = None
    revenue_threshold: Optional[Decimal] = None
    count_threshold: Optional[int] = None
    calculation_type: str = "percentage"
    value_numeric: Optional[Decimal] = None
    value_currency: Optional[Decimal] = None
    points_per_unit: Optional[Decimal] = None
    equity_units: Optional[Decimal] = None
    tier_table: Optional[List[Dict[str, Any]]] = None
    time_rate: Optional[Dict[str, Any]] = None
    cap_amount: Optional[Decimal] = None
    floor_amount: Optional[Decimal] = None
    payee_share_percent: Optional[Decimal] = None


class CommissionRuleUpdate(CommissionRuleCreate):
    pass


# ─── Assignment ───────────────────────────────────────────────────────────────

class CommissionAssignmentCreate(BaseModel):
    plan_id: UUID
    payee_id: UUID
    store_id: Optional[UUID] = None
    team_id: Optional[UUID] = None
    location: Optional[str] = None
    group_name: Optional[str] = None
    valid_from: Optional[date] = None
    valid_to: Optional[date] = None
    weight_percent: Decimal = Decimal("100")
    is_active: bool = True
    notes: Optional[str] = None


class CommissionAssignmentUpdate(BaseModel):
    store_id: Optional[UUID] = None
    team_id: Optional[UUID] = None
    location: Optional[str] = None
    group_name: Optional[str] = None
    valid_from: Optional[date] = None
    valid_to: Optional[date] = None
    weight_percent: Optional[Decimal] = None
    is_active: Optional[bool] = None
    notes: Optional[str] = None


# ─── Payout Run ───────────────────────────────────────────────────────────────

class PayoutRunCreate(BaseModel):
    period_start: Optional[date] = None
    period_end: Optional[date] = None
    payee_ids: Optional[List[str]] = None
    payment_method: str = "bank_transfer"
    notes: Optional[str] = None


class PayoutRunAction(BaseModel):
    notes: Optional[str] = None

File: vendor-web/src/types/commission.ts
// types/commission.ts

export type PayeeLinkType = 'vendor_user' | 'supplier' | 'customer' | 'external'
export type PayeeStatus = 'active' | 'inactive' | 'suspended'
export type PlanStatus = 'active' | 'inactive' | 'draft'
export type CalcType = 'percentage' | 'flat' | 'points' | 'tiered' | 'time_based' | 'revenue_based' | 'count_based' | 'equity'
export type AccrualStatus = 'draft' | 'accrued' | 'approved' | 'paid' | 'reversed' | 'disputed'
export type PayoutRunStatus = 'open' | 'approved' | 'paid' | 'cancelled'

export interface CommissionPayee {
  id: string
  vendor_id: string
  code?: string
  display_name: string
  phone?: string
  email?: string
  external_user_id?: string
  link_type: PayeeLinkType
  vendor_user_id?: string
  supplier_id?: string
  customer_id?: string
  default_payout_method: string
  currency: string
  status: PayeeStatus
  settings: Record<string, unknown>
  created_at?: string
  updated_at?: string
}

export interface TierRow { from: number; to?: number; rate?: number; flat?: number }

export interface CommissionRule {
  id: string
  plan_id: string
  vendor_id: string
  name?: string
  priority: number
  is_active: boolean
  applies_to: string
  product_id?: string
  service_id?: string
  category_id?: string
  uom?: string
  store_id?: string
  customer_group?: string
  channel: string
  event_tag?: string
  team_id?: string
  min_qty?: number
  min_amount?: number
  window_type: string
  period?: string
  revenue_threshold?: number
  count_threshold?: number
  calculation_type: CalcType
  value_numeric?: number
  value_currency?: number
  points_per_unit?: number
  equity_units?: number
  tier_table?: TierRow[]
  time_rate?: Record<string, unknown>
  cap_amount?: number
  floor_amount?: number
  payee_share_percent?: number
  created_at?: string
  updated_at?: string
}

export interface CommissionPlan {
  id: string
  vendor_id: string
  code: string
  name: string
  description?: string
  status: PlanStatus
  effective_from?: string
  effective_to?: string
  payee_scope: string
  priority: number
  stackable: boolean
  settings: Record<string, unknown>
  rules?: CommissionRule[]
  created_at?: string
  updated_at?: string
}

export interface CommissionAssignment {
  id: string
  vendor_id: string
  plan_id: string
  payee_id: string
  store_id?: string
  team_id?: string
  location?: string
  group_name?: string
  valid_from?: string
  valid_to?: string
  weight_percent: number
  is_active: boolean
  notes?: string
  created_at?: string
  /** Enriched by GET /assignments */
  payee_display_name?: string
  payee_link_type?: PayeeLinkType
  payee_email?: string
  payee_phone?: string
  plan_name?: string
  plan_code?: string
  employee_id?: string | null
}

export interface CommissionAccrual {
  id: string
  vendor_id: string
  payee_id: string
  plan_id?: string
  rule_id?: string
  assignment_id?: string
  source_type: string
  source_id: string
  source_line_ref?: string
  sale_date: string
  store_id?: string
  channel?: string
  base_amount: number
  calculation_type: CalcType
  value_applied?: number
  commission_amount: number
  points_amount: number
  equity_units_amount: number
  currency: string
  status: AccrualStatus
  payout_item_id?: string
  gl_entry_id?: string
  reversal_of?: string
  created_by_id?: string
  approved_by_id?: string
  approved_at?: string
  notes?: string
  created_at?: string
}

export interface CommissionPayoutItem {
  id: string
  run_id: string
  payee_id: string
  total_amount: number
  total_points: number
  total_equity: number
  accrual_count: number
  status: string
  payment_ref?: string
  paid_at?: string
  created_at?: string
}

export interface CommissionPayoutRun {
  id: string
  vendor_id: string
  run_no: string
  period_start?: string
  period_end?: string
  status: PayoutRunStatus
  total_amount: number
  total_points: number
  payee_count: number
  payment_method?: string
  gl_entry_id?: string
  notes?: string
  created_by_id?: string
  approved_by_id?: string
  approved_at?: string
  paid_at?: string
  created_at?: string
  items?: CommissionPayoutItem[]
}

export interface CommissionSummary {
  total_accrued: number
  total_paid: number
  pending_approval: number
  avg_per_sale: number
  top_payee_id?: string
  top_payee_amount: number
  sale_count: number
}

export interface ByPayeeRow {
  payee_id: string
  payee_name: string
  total_commission: number
  total_base: number
  count: number
}

export interface TrendRow { period: string; total: number; count: number }

export interface BySourceResult {
  by_channel: { channel: string; total: number; count: number }[]
  by_source_type: { source_type: string; total: number; count: number }[]
}

export interface PaginatedResult<T> {
  items: T[]
  total: number
  page: number
  size: number
  pages: number
}
